KEP 5339 - ClusterProfile Credentials external providers (#5338)
kubernetes/enhancements: keps/sig-multicluster/5339-clusterprofile-plugin-credentials/README.md

that want to operate on the clusters. However, there isn't a single way to obtain credentials
to reach those clusters. This KEP provides a standardized way to obtain credentials for Clusters
when using ClusterProfile and makes it pluggable to allow the diverse ecosystem to support the
multitude of ways to obtain credentials. It also reuses part of the Kubeconfig external provider
semantics to make implementation easier.

## Motivation

ClusterInventory is unfinished without an ability to use the clusters and controller
writers have been very explicit that credentials are needed. Previous attempts at writing credentials have
failed and we believe that a plugin model, also reusing known flows, will help solve the "credentials" need
for ClusterProfiles.

See [introduction slides](https://docs.google.com/presentation/d/1v5-J-kFJ3TSpKqSraHcYkCz2NG7cNnYpq0ISF85wNMU/edit)

### Goals

* Provide a library for controllers to obtain credentials for a cluster represented by a ClusterProfile
* Allow cluster managers to provide a method to obtain credentials that doesn't require to be embedded into the controller code
  and recompiling.

> * Be a secure mechanism for credential obtention and storage.

Reviewer: I guess it would not hurt to get a buy-in from the sig Auth

Author: I agree, could you invite a couple people to this KEP? I don't remember POCs.
Given how similar it is to the kubeconfig exec, I do expect good inputs from them.

### Non-Goals

* Define the mechanism for shipping plugins to be used by the controllers
* Design plugin or a library for plugins
* Mandate Federated workload identity / OIDC frameworks (though they are recommended)

## Proposal

The proposed approach to issuing credentials is to leverage plugins for credentials, where a controller could leverage a library running local code
which would retrieve the credentials for the current controller and for the clusterprofile of their choice. It is expected that plugins would

> leverage the identity of the controller (for example, the Kubernetes Service Account when running in Kubernetes) to retrieve credentials that are valid on other clusters.

Reviewer 2: The KEP does not seem to directly mention using the controller's identity anywhere else but in this paragraph. I am not sure workload identity is currently typical for use with exec plugins but I may be wrong.

Plugins would be exec'ed by the controller so that they don't need be built-in the binary, allowing cluster managers to
write their own credentials and still leveraging multicluster controllers written by the community.
In addition, we propose to reuse the exec approach and protocol used for external credentials in Kubeconfig. Finally,
in order to retrieve the endpoint for the cluster, we standardize the property names that are used in ClusterProfile.

> ### Risks and Mitigations

Author: added content, ptal

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

> The proposal's implementation would be done via a Library in https://github.com/kubernetes-sigs/cluster-inventory-api.

Author: from what I saw, we don't have a controller-runtime dependency, only client go and exec of cmdapi.

The library would be in golang. The library is provided as the community shared implementation for golang and
it is possible that other implementations would be created, and would work with the same plugin mechanism defined here,
allowing for reuse of the external providers that cluster managers write.

The expected prototype for a controller is expected to be the following:

`func (c *ClusterProfileExternalProviders) GetConfig(cp ClusterProfile) (rest.Config, error)`

The library implementation flow is expected to be as follow:

1. Build the endpoint details of the cluster by reading properties of the ClusterProfile

> 2. Call the CredentialsExternalProviders, following the same flow defined in [KEP 541](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/541-external-credential-providers/README.md)

Author: the kubeconfig would be unnecessary, it would likely add complexity (writing then reading the kubeconfig) as well as delay the plugin call (since it would be configured in the kubeconfig), which would likely make for a worse error handling.

Given how easy it is to call exec with the same syntax (shown in https://github.com/kubernetes-sigs/cluster-inventory-api/pull/17); I think we can call the exec code without relying on the full kubeconfig.

3. Build the rest.Config and return it to the caller

### Converting endpoint properties to cluster object

> The Cluster structure for the exec defined in KEP 541 assumes the following:

Reviewer: I think we need to have an additional free form "data" in the clusterProfile API so that each "plugin" can utilize during the authentication process. i.e. Azure plugin would need the cluster's tenant and subscription ID.

```

> type Cluster struct {

Reviewer 2: My bad, this v1 definition is probably more suiting https://github.com/kubernetes/kubernetes/blob/a34c07971b610eb33908a743eadb4c61beeecc50/staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1/types.go#L78-L79

  // extensions[client.authentication.k8s.io/exec] field:
  //
  // clusters:
  // - name: my-cluster
  //   cluster:
  //     ...
  //     extensions:
  //     - name: client.authentication.k8s.io/exec  # reserved extension name for per cluster exec config
  //       extension:
  //         audience: 06e3fbd18de8  # arbitrary config
  //
  // In some environments, the user config may be exactly the same across many clusters
  // (i.e. call this exec plugin) minus some details that are specific to each cluster
  // such as the audience.  This field allows the per cluster config to be directly
  // specified with the cluster info.  Using this field to store secret data is not
  // recommended as one of the prime benefits of exec plugins is that no secrets need
  // to be stored directly in the kubeconfig.
  // +optional
  Config runtime.RawExtension `json:"config,omitempty"`
}
```

This data must be created from the ClusterProfile, as described in the following section.

### Standardizing the Endpoint Property

In order to populate the Cluster object that the exec provider requires, we standardize the following properties.

> Property: `multicluster.sigs.io/execprovider-cluster`

Reviewer 3: Hi corentone! This information would be exposed to all readers of the Cluster Profile object, which I fear might incur some security concerns?

Reviewer 3: Another part is that different plugins might require different cluster-specific information -> would this be a free-form property?

Reviewer 3: Each plugin would also need some application specific information to make the authentication work. Those obviously are not ideal to be exposed directly on Cluster Profile; I am wondering if we could/should formulate a way where application specific information can be discovered as well.

Reviewer 3: As for the cluster-specific and application-specific data, we might have gone over this brief on the Slack channel; basically, depending on how the authn workflow is implemented (there are obviously a variety of options), each plugin would need its own set of inputs.

For example:

* for a plugin that, let's just assume can magically (and securely) acquires a service account token out of thin air, it would need the following items to complete the auth flow:
    * (cluster-specific info.): API server address and CA bundle
    * (application-specific info.): the service account name that represents the application (that invokes the plugin)

* for a plugin that uses federated identities, it would need:
    * (cluster-specific info.):  API server address, CA bundle, expected audience for token exchange, and some platform specific information (on AKS -> tenant ID, on GKE -> workload identity provider full ID)
    * (application-specific info.): (on AKS -> client ID)

* for a plugin that uses SPIRE, it would need:
    * (cluster-specific info.): API server address, CA bundle, expected audience
    * (application-specific info.): spiffe ID

Considering flexibility reasons, we might need to use something free-form I assume? Of course we could just assume that everything is read from local setup (mounted volumes, configMaps, etc.). but it kinds of defeats the benefits of having cluster profiles (esp. for the cluster-specific information part). If we could have application-specific information linked with Cluster Profiles objects (this one must be done via object references, for obvious reasons), it could further simplify the flow I guess.

Reviewer 3: This might be more relevant to another comment I posted, but I am wondering if we could have something like (modifying the Cluster Profile API):

```yaml
...
spec:
    authProviders:
    - type: aksoidc
       clusterInfo:
       # we could do either inline (arbitrary map)...
           caData: ...
           server: ...
           audience: ...
           tenantID: ...
       # or use object refs
           objectRef: ....
       targetApplications:
       - name: argo
          applicationInfo:
              objectRef: ...
       - name: multi-kueue
          applicationInfo:
              objectRef: ...
```

This way each type of plugins could work (`exec` typed or libraries, not suggesting that this KEP should touch the latter, though), and it saves the trouble of users having to mount a bunch of information manually just for auth purposes, esp. considering that the info would be cluster-specific.

Author: **application specific info are provided by the controller when setting up the plugin, never by cluster profile.** ClusterProfile is not aware of the applications using it. Cluster profile says what "plugin types" work with it.
Now yes the application will have to provide that data to the plugin. That will likely be done via volume, configmaps, envvars etc. Given the plugin is exec'ed, it has a access to the same data as the controller itself (volumes/FS, environment, KSA...).
IMO having application-specific information is part of what doomed the previous KEP on credentials and we can't make that mistake again. It would require ClusterProfile to know about all the applications that can use it and how they are setup. I don't see that approach working.

The clusterprofile needs to provide all the cluster-specific information (endpoints and connectivity as well as plugin types and pointers to what authN mechanism is recommended).

Let me do another pass at the KEP to provide a new ClusterProfile example as well as a Controller example.
Fwiw, for GKE, it would leverage the federated identity, which can be obtained just by calling the metadata server, so clusterprofile doesn't need to provide us with anything.

Reviewer 3: Having the application specific info via object refs could help simplify the process, but I would understand if it's something that we want to leave for the application itself to set up, i.e., taking an opinionated approach on the API part and making sure that any sensitive info cannot be added). To me personally, the reason why the last push wasn't successful is mostly that we were trying to encourage a practice that has security concerns (even though it is commonly used).

Looking forward to seeing the new pass. It would be really nice if we could:

* on the API level having neutral names so that no limits are implicitly imposed on how the flow is implemented (of course for the KEP, and the current efforts in the WG, I totally understand that we are taking a `exec` plugin centric approach)
* we would have a field that allows placement of arbitrary cluster-specific authentication info as different authn workflows might have different requirements (though some fields might be very commonly used). I fear that not all platforms would handle the flow as GKE does (metadata server), especially for those who have a hybrid environment.

> Property: `multicluster.sigs.io/execprovider-cluster`

Reviewer 4: Why `execprovider` (which feels implementation-derived) rather than something aligned with the feature provided, such as `credprovider`?

Reviewer 4: Ah, is the intent to match the `exec` section in `kubeconfig`?

Value: json containing the Cluster struct (see above).
Example Value: "{\"Server\":\"https://example.com:443\"}"

### Configuring plugins in the controller

> Plugins are designated by a hardcoded string, for example, "gke" for GKE Clusters, which allows the controller

Reviewer 2: nit, maybe?
```suggestion
Plugins are selected by a hardcoded string, for example, "gke" for GKE Clusters, which allows the controller
```

to attach a different binary name or path for the the binary. It is expected that the library will have a mapping from

> its supported plugin names to the expected binary to call. It is expected that the mapping is provided via a flag on the controller.

Reviewer 2: Wouldn't mapping binaries make this vendor/implementation specific? It's probably not useful to just hardcode the binary if you don't know how it's configured, right? I.e. hardcoding a binary only specifies the `Command` part of the [ `ExecConfig`](https://github.com/kubernetes/kubernetes/blob/a34c07971b610eb33908a743eadb4c61beeecc50/staging/src/k8s.io/client-go/tools/clientcmd/api/types.go#L206-L208).

Author: I added clarification between the string and the plugin.

The string to map to plugin is the "credentials type", then the controller has to setup a plugin for that "credentials type".

A credentials Type could be: `google`, `azure`, `direct` and then the controller would have to plug the right plugin (there may be multiple that works for a given credentials).

For google, I'm guessing one could use workload identity and therefore use a plugin called `google-with-wi` which would read local files/env to use WI that is a google creds. That same, google identity could map to a `google-user` which would use an interactive login?

```
./controller ... --clusterprofile-creds-providers "gke=gke-gcloud-auth-plugin"
```

### Selecting the right plugin

> The plugin is designated by the clusterprofile, which designates the type of provider that is expected to be called.

Reviewer 2: ```suggestion
The plugin is specified in the clusterprofile, which designates the type of provider that is expected to be called.
```

Author: applied

The clusterprofile has a property to designate the plugin.

> Property: `multicluster.sigs.io/execprovider-name`

Reviewer 3: Hi corentone! I assume that here we implicitly require that one cluster could have only one provider?

Author: yes but the controller could decide what plugin to use. Which also means they could use a different binary for a given provider name.

For ex, the cluster could be GKE, but as a controller I could use the official-gke-fleet plugin or official-gke-as-user plugin or i could write my own: gke-custom-plugin mechanism.
Any cluster seen as GKE would use the chosen plugin.

Value: string that is expected from the controller to satisfy via plugins.
Example Value: "gke"

### External credentials Provider plugin mechanism

In order to call the plugin, the library execs the plugin defined in the configuration. It passes the Cluster information that

> was obtained from the ClusterProfile.

Reviewer: this is where we need an extra field/status in the ClusterProfile.  We can also make this as a standard property if a field feels a bit too rigid.

Author: moved to a field.

> The library then calls the plugin following the protocol defined in [KEP 541](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/541-external-credential-providers/README.md). The library provided in https://github.com/kubernetes-sigs/cluster-inventory-api can leverage the original code that is kept in

Reviewer 4: `clientcmd`?

Author: yes I think thats the one. I may have left this sentence hanging looking for the code. oops.
will add it!

Author: fixed

Consider the following in developing the graduation criteria for this enhancement:
- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
- [Feature gate][feature gate] lifecycle
- [Deprecation policy][deprecation-policy]

Clearly define what graduation means by either linking to the [API doc
definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning)
or by redefining what graduation means.

In general we try to use the same stages (alpha, beta, GA), regardless of how the
functionality is accessed.

[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/

Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].

#### Alpha

- Feature implemented behind a feature flag
- Initial e2e tests completed and enabled

#### Beta

- Gather feedback from developers and surveys
- Complete features A, B, C

> - Additional tests are in Testgrid and linked in KEP

Reviewer: what is a Testgrid?

Author: I didn't touch the template. but at google testgrid is a grid with all test results across date/testruns.

Author: resolving to remove clutter. we can cleanup all commented code right before merging.

In order to call the plugin, the library execs the plugin defined in the configuration. It passes the Cluster information that was obtained from the ClusterProfile.
The library then calls the plugin following the protocol defined in [KEP 541](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/541-external-credential-providers/README.md).
The library provided in https://github.com/kubernetes-sigs/cluster-inventory-api can leverage the [original code that is kept in client-go](https://github.com/kubernetes/client-go/blob/master/plugin/pkg/client/auth/exec/exec.go#L159).

### Standardizing the Provider definition

In order to populate the Cluster object that the exec provider requires, we standardize a new field in ClusterProfile called `credentials` that is stored in the Spec of the ClusterProfile.
All the data from this structure is specific to the clusterProfile and does not contain any Controller-specific information. It must be usable by different
controller, applications or consumers without requiring changes. It also cannot contain any data considered a secret; and we consider that reachability information
is not sensitive.

The definition is as follows:
```

> type Credentials struct {

Author: done, renamed CredentialProviders because we already had a CredentialsConfig

```
./controller ... --clusterprofile-creds-provider "google='/usr/bin/gke-gcloud-auth-plugin --flag1 value1 --flag2 value2'"
```

Despite being a flag, we can express the equivalent structure for each Plugin:
```
type Provider struct {
  CredentialsType string
  ExecutablePath string
  args []string
}
```

Given the plugin is executed directly by the controller, it may expect to have access to the same environment as the controller itself, inclusive of envvars, filesystem and network.
It is expected that the identity of the plugin is the same as the controller itself.

### Plugin Examples

As an example, we provide pseudocode for plugins that could easily be implemented with the protocol. They are ultrasimplified
version of the code and structures to convey the idea and not be an implementation example.

#### Secret Reader plugin

This plugin assumes the controller is aware of the list of clusters ahead of time and has created secrets for them.
It simply reads the token from the secret mapped to the cluster.

```

> func GetToken(clusterName string) string {

Reviewer: I wonder how we support creating different identities for different applications (i.e.  argoCD vs Kueue) when the input is only clusterName? Where can the library know which app it is generating the token for?

Author: yes, valid concern which means this example is not complete enough.
Let me flush out a bit more details. The goal is to show how someone could do it and it's true we shouldn't encourage bad behaviors.

Author: flushed a bit.

that want to operate on the clusters. However, there isn't a single way to obtain credentials
to reach those clusters. This KEP provides a standardized way to obtain credentials for Clusters
when using ClusterProfile and makes it pluggable to allow the diverse ecosystem to support the
multitude of ways to obtain credentials. It also reuses part of the Kubeconfig external provider
semantics to make implementation easier.

## Motivation

ClusterInventory is unfinished without an ability to use the clusters and controller
writers have been very explicit that credentials are needed. Previous attempts at writing credentials have
failed and we believe that a plugin model, also reusing known flows, will help solve the "credentials" need
for ClusterProfiles.

See [introduction slides](https://docs.google.com/presentation/d/1v5-J-kFJ3TSpKqSraHcYkCz2NG7cNnYpq0ISF85wNMU/edit)

### Goals

* Provide a library for controllers to obtain credentials for a cluster represented by a ClusterProfile
* Allow cluster managers to provide a method to obtain credentials that doesn't require to be embedded into the controller code

>   and recompiling.

Author: Imagine you are a controller writer and you want to share your controller, for example ArgoCD or MultiKueue.
You don't want to compile the controller for ALL providers. You want to ship a single docker image (or binary version/package).

If recompilation was required, that means that any divergence in credentials provider would require to fork, edit and rebuild any controller you want to use. I know for me, I would definitely not want having to recompile argocd to install it on my GKE Fleet. Nor I don't want to make PRs to add GKE support into argocd.

	// +optional
	DisableCompression bool `json:"disable-compression,omitempty"`
	// Config holds additional config data that is specific to the exec
	// plugin with regards to the cluster being authenticated to.
	//
	// This data is sourced from the clientcmd Cluster object's
	// extensions[client.authentication.k8s.io/exec] field:
	//
	// clusters:
	// - name: my-cluster
	//   cluster:
	//     ...
	//     extensions:
	//     - name: client.authentication.k8s.io/exec  # reserved extension name for per cluster exec config
	//       extension:
	//         audience: 06e3fbd18de8  # arbitrary config
	//
	// In some environments, the user config may be exactly the same across many clusters
	// (i.e. call this exec plugin) minus some details that are specific to each cluster
	// such as the audience.  This field allows the per cluster config to be directly
	// specified with the cluster info.  Using this field to store secret data is not
	// recommended as one of the prime benefits of exec plugins is that no secrets need
	// to be stored directly in the kubeconfig.
	// +optional
	Config runtime.RawExtension `json:"config,omitempty"`

> }```

Reviewer 4: The three backticks need to be on their own line (the markup is broken in the rest of the file as a result).

	// attempts to construct a proxy configuration from http_proxy and
	// https_proxy environment variables. If these environment variables are not
	// set, the client does not attempt to proxy requests.
	//
	// socks5 proxying does not currently support spdy streaming endpoints (exec,
	// attach, port forward).
	// +optional
	ProxyURL string `json:"proxy-url,omitempty"`
	// DisableCompression allows client to opt-out of response compression for all requests to the server. This is useful
	// to speed up requests (specifically lists) when client-server network bandwidth is ample, by saving time on
	// compression (server-side) and decompression (client-side): https://github.com/kubernetes/kubernetes/issues/112296.
	// +optional
	DisableCompression bool `json:"disable-compression,omitempty"`
	// Extensions holds additional information. This is useful for extenders so that reads and writes don't clobber unknown fields
	// +optional
	Extensions map[string]runtime.Object `json:"extensions,omitempty"`
}
```

In this structure, not all fields would apply, such as:

* `CertificateAuthority`, which points to a file (and a ClusterProfile doesn't have a filesystem)

> * `Extensions`, while extensions may be good in the future, we may not leverage them.

Reviewer 3: Actually this seems like a perfect place for cluster-specific info. if the plugin type needs it. Could we say that it is up to the user/plugin type to decide whether it can be leveraged (I know it's already implied, but this line seems to suggest that it is reserved only for future uses).

writers have been very explicit that credentials are needed. Previous attempts at writing credentials have
failed and we believe that a plugin model, also reusing known flows, will help solve the "credentials" need
for ClusterProfiles.

See [introduction slides](https://docs.google.com/presentation/d/1v5-J-kFJ3TSpKqSraHcYkCz2NG7cNnYpq0ISF85wNMU/edit)

### Goals

* Provide a library for controllers to obtain credentials for a cluster represented by a ClusterProfile
* Allow cluster managers to provide a method to obtain credentials that doesn't require to be embedded into the controller code
  and recompiling.
* Be a secure mechanism for credential obtention and storage.

### Non-Goals

* Define the mechanism for shipping plugins to be used by the controllers and their delivery in the controller image/pod.
* Design plugin or a library for plugins
* Mandate Federated workload identity / OIDC frameworks (though they are recommended)

## Proposal

The proposed approach to obtaining credentials is to leverage plugins for retrieving the credentials from an issuer recognized by the target cluster. The controller using ClusterProfile

> would use a library running a local executable which would retrieve the credentials for the current controller and a given clusterprofile.

Author: done

It is expected that plugins would leverage elements local to the controller to help assert the identity of the controller (environment variables, config files, KSA, the local IP, etc...)

> to retrieve credentials that are valid on the target cluster. Plugins would be exec'ed by the controller so that they don't need be built-in the binary, allowing flexibility into writing their

Author: I agree with the risk you explain. This is what I tried to call out.
Let me try to dissect your suggestion. It would look like:

`Controller (1) -> Local HTTP Endpoint For creds services (2) -> Call exec locally to retrieve creds (3)`

Would this service that retrieves the cred (2+3) be dedicated for that controller or shared? How would it make sure it is handling credentials for the right pod?
I think you're describing the design of MDS which deals with credentials in GKE. It is a well known IP that is used to issue credentials. It identifies pods and retrieves google credentials for them.

The problem is that it's a lot more complicated to write and deploy I think. We could technically later replace our exec mechanism with this flow? Also that means you trust this central controller to issue the credentials instead of having each controller do its own exchange.

> to retrieve credentials that are valid on the target cluster. Plugins would be exec'ed by the controller so that they don't need be built-in the binary, allowing flexibility into writing their

Author: given your approval, Im guessing it does, resolving.

	// extensions[client.authentication.k8s.io/exec] field:
	//
	// clusters:
	// - name: my-cluster
	//   cluster:
	//     ...
	//     extensions:
	//     - name: client.authentication.k8s.io/exec  # reserved extension name for per cluster exec config
	//       extension:
	//         audience: 06e3fbd18de8  # arbitrary config
	//
	// In some environments, the user config may be exactly the same across many clusters
	// (i.e. call this exec plugin) minus some details that are specific to each cluster
	// such as the audience.  This field allows the per cluster config to be directly
	// specified with the cluster info.  Using this field to store secret data is not
	// recommended as one of the prime benefits of exec plugins is that no secrets need
	// to be stored directly in the kubeconfig.
	// +optional
	Config runtime.Object
}
```

#### ClusterProfile Example

Example of a GKE ClusterProfile, which would map to a plugin providing credentials of type `google`:
```
apiVersion: multicluster.x-k8s.io/v1alpha1
kind: ClusterProfile
metadata:
 name: my-cluster-1
spec:
  displayName: my-cluster-1
  clusterManager:
    name: GKE-Fleet

>   credentials:

Author: if you prefer, I can move it; I'll ask for confirmation tomorrow.

>   credentials:

Author: moved, lets see if people scream :)

    google:
      cluster:
        server: https://connectgateway.googleapis.com/v1/projects/123456789/locations/us-central1/gkeMemberships/my-cluster-1
status:
  version:
    kubernetes: 1.28.0
  properties:
   - name: clusterset.k8s.io
     value: some-clusterset
   - name: location
     value: us-central1
```

### Configuring plugins in the controller

Plugins are selected by a hardcoded string which represents the type of credentials that is expected by the cluster, for example, "google" for GKE Clusters.

> Thish allows the controller to attach a different binary name or path for the the binary.

Reviewer 5: nit: s/Thish/This/
s/the the/the/

The definition is as follows:
```
type Credentials struct {
  // +listType=map
  // +listMapKey=name

>   credentialProviders []CredentialsConfig // mapping of credentials types to their config. In some cases the cluster may recognize different identity types and they may have different endpoints or TLS config.

Reviewer 6: Since in clusterProfile spec, there is clusterManager field which means a clusterPofile can only be owned by a single clusterManager. Do we still need this to be a list here, such that controller has to select one of the config from it. Or it is clusterManager's job should be responsible to provide a single config?

Author: I think even if its a single ClusterManager, there could be multiple paths to reach the cluster.
For example, a GKE cluster could be accessed via GKE Connect Gateway (the fleet endpoint) or via GKFE (the normal endpoint). Another example would be a cluster that is accessible from onprem (but requires proxy and special login etc etc) and another path via a proxy that does special checks. Also in case of transitions... a platform admin may want to offer alternative endpoints and offer a way to slowly migrate.

In general it doesn't cost us much to support a list and I think it adds flexibility.

  Name string // name of the provider type
  Cluster *Cluster // Configuration to reach the cluster (endpoints, proxy, etc) // See following section for details.
}
```

#### Cluster Data

The Cluster structure for the exec defined in KEP 541, [implemented in k/k](https://github.com/kubernetes/kubernetes/blob/a34c07971b610eb33908a743eadb4c61beeecc50/staging/src/k8s.io/client-go/pkg/apis/clientauthentication/types.go#L73-L80) assumes the following:

```
type Cluster struct {
	// Server is the address of the kubernetes cluster (https://hostname:port).
	Server string `json:"server"`
	// TLSServerName is passed to the server for SNI and is used in the client to
	// check server certificates against. If ServerName is empty, the hostname
	// used to contact the server is used.
	// +optional
	TLSServerName string `json:"tls-server-name,omitempty"`
	// InsecureSkipTLSVerify skips the validity check for the server's certificate.
	// This will make your HTTPS connections insecure.
	// +optional
	InsecureSkipTLSVerify bool `json:"insecure-skip-tls-verify,omitempty"`
	// CAData contains PEM-encoded certificate authority certificates.
	// If empty, system roots should be used.

> 	// +listType=atomic

Reviewer 6: We cannot put this into CRD because the listType setting here.

Author: why?
```
// Example:
//
//	+listType=atomic
//
// Atomic lists will be entirely replaced when updated. This tag may be
// used on any type of list (struct, scalar, ...).
//
// Using this tag will generate the following OpenAPI extension:
//
//	"x-kubernetes-list-type": "atomic"
```

whats wrong with this? I think it applies only to `CertificateAuthorityData` field.

Author: ah bummer. maybe we could check with the original library, this may be a miss on their end?

Reviewer 4: See the comment in the original file:

> To ensure that this struct contains everything someone would need to communicate
> with a kubernetes cluster (just like they would via a kubeconfig), the fields
> should shadow "k8s.io/client-go/tools/clientcmd/api/v1".Cluster, with the exception
> of CertificateAuthority, since CA data will always be passed to the plugin as bytes.

The data type exported to clients is https://github.com/kubernetes/kubernetes/blob/a34c07971b610eb33908a743eadb4c61beeecc50/staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1/types.go#L78 which doesn’t have `CertificateAuthorityData`.
